core/tee/tadb.c

```diff
@@ -0,0 +1,702 @@
+	tadb_put(ta->db);
+	free(ta);
+	return TEE_SUCCESS;
+
+err:
+	tee_tadb_ta_close_and_delete(ta);
+	return res;
+}
+
+TEE_Result tee_tadb_ta_delete(const TEE_UUID *uuid)
+{
+	const struct tadb_entry null_entry = { 0 };
+	struct tee_tadb_dir *db;
+	struct tadb_entry entry;
+	size_t idx;
+	TEE_Result res;
+
+	res = tee_tadb_open(&db);
+	if (res)
+		return res;
+
+	res = find_ent(db, uuid, &idx, &entry);
+	if (res) {
+		tee_tadb_close(db);
+		return res;
+	}
+
+	clear_file(db, entry.file_number);
+	res = write_ent(db, idx, &null_entry);
+	tee_tadb_close(db);
+	if (res)
+		return res;
+
+	ta_operation_remove(entry.file_number);
+	return TEE_SUCCESS;
+}
+
+TEE_Result tee_tadb_ta_open(const TEE_UUID *uuid, struct tee_tadb_ta **ta_ret)
+{
+	TEE_Result res;
+	size_t idx;
+	struct tee_tadb_ta *ta = calloc(1, sizeof(*ta));
+	static struct tadb_entry last_entry;
+
+	if (!ta)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	if (!memcmp(uuid, &last_entry.prop.uuid, sizeof(*uuid))) {
+		ta->entry = last_entry;
+	} else {
+		res = tee_tadb_open(&ta->db);
+		if (res)
+			goto err;
+
+		res = find_ent(ta->db, uuid, &idx, &ta->entry);
+		if (res)
+			goto err;
+	}
+
+	res = ta_operation_open(OPTEE_MRF_OPEN, ta->entry.file_number, &ta->fd);
+	if (res)
+		goto err;
+
+	res = tadb_authenc_init(TEE_MODE_DECRYPT, &ta->entry, &ta->ctx);
+	if (res)
+		goto err;
+
+	*ta_ret = ta;
+
+	return TEE_SUCCESS;
+err:
+	tadb_put(ta->db);
+	free(ta);
+	return res;
+}
+
+const struct tee_tadb_property *tee_tadb_ta_get_property(struct tee_tadb_ta *ta)
+{
+	return &ta->entry.prop;
+}
+
+static TEE_Result ta_load(struct tee_tadb_ta *ta)
+{
+	TEE_Result res;
+	const size_t sz = ta->entry.prop.custom_size + ta->entry.prop.bin_size;
+	struct optee_msg_param params[2] = {
+		[0] = { .attr = OPTEE_MSG_ATTR_TYPE_VALUE_INPUT,
+			.u.value.a = OPTEE_MRF_READ,
+			.u.value.b = ta->fd,
+			.u.value.c = 0 },
+	};
+
+	if (ta->ta_mobj)
+		return TEE_SUCCESS;
+
+	ta->ta_mobj = thread_rpc_alloc_payload(sz, &ta->ta_cookie);
+	if (!ta->ta_mobj)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	ta->ta_buf = mobj_get_va(ta->ta_mobj, 0);
+	assert(ta->ta_buf);
+
+	if (!msg_param_init_memparam(params + 1, ta->ta_mobj, 0, sz,
+				     ta->ta_cookie, MSG_PARAM_MEM_DIR_OUT))
+		return TEE_ERROR_BAD_STATE;
+
+	res = thread_rpc_cmd(OPTEE_MSG_RPC_CMD_FS, ARRAY_SIZE(params), params);
+	if (res) {
+		thread_rpc_free_payload(ta->ta_cookie, ta->ta_mobj);
+		ta->ta_mobj = NULL;
+	}
+	return res;
+}
+
+TEE_Result tee_tadb_ta_read(struct tee_tadb_ta *ta, void *buf, size_t *len)
+{
+	TEE_Result res;
+	const size_t sz = ta->entry.prop.custom_size + ta->entry.prop.bin_size;
+	size_t l = MIN(*len, sz - ta->pos);
+
+	assert(tadb_check_state(ta, TEE_MODE_DECRYPT));
+
+	res = ta_load(ta);
+	if (res)
+		return res;
+
+	if (buf) {
+		res = tadb_update_payload(ta->ctx, TEE_MODE_DECRYPT,
+					  ta->ta_buf + ta->pos, l, buf);
+		if (res)
+			return res;
+	} else {
+		size_t num_bytes = 0;
+
+		while (num_bytes < l) {
+			uint8_t b[1024 * 4];
```

Comment:
Are you really sure, you want  to allocate 4kb on stack?

Reply:
You're right, it's a bit excessive. I'll replace it with `malloc()`.